Brokerages/Binance/BinanceBrokerage.Messaging.cs

```diff
@@ -0,0 +1,317 @@
+                    Symbol = symbol,
+                    TickType = TickType.Quote,
+                    AskSize = askSize,
+                    BidSize = bidSize
+                });
+            }
+        }
+
+        private void EmitTradeTick(Symbol symbol, DateTime time, decimal price, decimal quantity)
+        {
+            lock (TickLocker)
+            {
+                Ticks.Add(new Tick
+                {
+                    Symbol = symbol,
+                    Value = price,
+                    Quantity = Math.Abs(quantity),
+                    Time = time,
+                    TickType = TickType.Trade
+                });
+            }
+        }
+
+        private void OnFillOrder(Execution data)
+        {
+            try
+            {
+                var order = FindOrderByExternalId(data.OrderId);
+                if (order == null)
+                {
+                    // not our order, nothing else to do here
+                    return;
+                }
+
+                var symbol = _symbolMapper.GetLeanSymbol(data.Symbol);
+                var fillPrice = data.LastExecutedPrice;
+                var fillQuantity = data.LastExecutedQuantity;
+                var updTime = Time.UnixMillisecondTimeStampToDateTime(data.TransactionTime);
+                var orderFee = new OrderFee(new CashAmount(data.Fee, data.FeeCurrency));
+                var status = ConvertOrderStatus(data.OrderStatus);
+                var orderEvent = new OrderEvent
+                (
+                    order.Id, symbol, updTime, status,
+                    data.Direction, fillPrice, fillQuantity,
+                    orderFee, $"Binance Order Event {data.Direction}"
+                );
+
+                if (status == OrderStatus.Filled)
+                {
+                    Orders.Order outOrder;
+                    CachedOrderIDs.TryRemove(order.Id, out outOrder);
+                }
+
+                OnOrderEvent(orderEvent);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e);
+                throw;
+            }
+        }
+
+        private Orders.Order FindOrderByExternalId(string brokerId)
+        {
+            var order = CachedOrderIDs
+                    .FirstOrDefault(o => o.Value.BrokerId.Contains(brokerId))
+                    .Value;
+            if (order == null)
+            {
+                order = _algorithm.Transactions.GetOrderByBrokerageId(brokerId);
+            }
+
+            return order;
+        }
+
+        private void FetchOrderBookSnapshot(Symbol symbol, BinanceOrderBook orderBook)
+        {
+            LockStream();
+
+            var endpoint = $"/api/v1/depth?symbol={_symbolMapper.GetBrokerageSymbol(symbol)}&limit=1000";
+            var request = new RestRequest(endpoint, Method.GET);
+            var response = ExecuteRestRequest(request);
+
+            if (response.StatusCode != HttpStatusCode.OK)
+            {
```

Comment:
Looks like the stream doesn't get unlocked in this code path. When doing lock/unlock semantics, it's best to have it be done using a `try/finally` block to guarantee we always call `UnlockStream`. Another option is to encapsulate the locking in a single function:
``` csharp
private void WithLockedStream(Action code)
{
    try
    {
        LockStream();
        code();
    }
    finally
    {
        UnlockStream();
    }
}
```
and then consumers can call via:
``` csharp
WithLockedStream(() =>
{
    // code that requires the stream to be locked
});
```

Reply:
Also, it's most likely the case that we don't need to lock the stream over the _entire_ method body, but only the sections that are sensitive to the stream being locked.